pkg/operator/encryption/state.go

```diff
@@ -43,175 +45,241 @@ func getCurrentEncryptionConfig(secrets corev1client.SecretInterface, revision s
+	//
+	// STEP 4: with consistent read-keys and write-keys, remove every read-key other than the write-key.
+	//
+	// Note: because read-keys are consistent, currentlyEncryptedGRs equals toBeEncryptedGRs
+	allMigrated, _, reason := migratedFor(currentlyEncryptedGRs, writeSecret)
+	if !allMigrated {
+		klog.V(4).Infof(reason)
+		return desiredEncryptionState
+	}
+	for gr := range desiredEncryptionState {
+		grState := desiredEncryptionState[gr]
+		grState.readSecrets = []*corev1.Secret{writeSecret}
+		desiredEncryptionState[gr] = grState
+	}
+	klog.V(4).Infof("write secret %s set as sole write key", writeSecret.Name)
+	return desiredEncryptionState
+}
 
-	migratedResourceString := writeSecret.Annotations[encryptionSecretMigratedResources]
-	// if no migration has happened, we need to wait until it has
+// migratedFor returns whether all given resources are marked as migrated in the given secret.
+// It returns missing GRs and a reason if that's not the case.
+func migratedFor(grs []schema.GroupResource, s *corev1.Secret) (ok bool, missing []schema.GroupResource, reason string) {
+	migratedResourceString := s.Annotations[encryptionSecretMigratedResources]
 	if len(migratedResourceString) == 0 {
-		klog.V(4).Infof("write secret %s has not been migrated", writeSecret.Name)
-		return encryptionState
+		return false, grs, fmt.Sprintf("secret %s has not been migrated", s.Name)
 	}
-	migratedResources := &migratedGroupResources{}
-	// if we cannot read the data, wait until we can
-	if err := json.Unmarshal([]byte(migratedResourceString), migratedResources); err != nil {
-		klog.V(4).Infof("write secret %s has invalid migrated resource string: %v", writeSecret.Name, err)
-		return encryptionState
+
+	migrated := &migratedGroupResources{}
+	if err := json.Unmarshal([]byte(migratedResourceString), migrated); err != nil {
+		return false, grs, fmt.Sprintf("secret %s has invalid migrated resource string: %v", s.Name, err)
 	}
 
-	// get a list of all the resources we have, so that we can compare against the migrated keys annotation.
-	allResources := make([]schema.GroupResource, 0, len(encryptionState))
-	for gr := range encryptionState {
-		allResources = append(allResources, gr)
+	var missingStrings []string
+	for _, gr := range grs {
+		if !migrated.hasResource(gr) {
+			missing = append(missing, gr)
+			missingStrings = append(missingStrings, gr.String())
+		}
 	}
 
-	foundCount := 0
-	for _, expected := range allResources {
-		if migratedResources.hasResource(expected) {
-			foundCount++
+	if len(missing) > 0 {
+		return false, missing, fmt.Sprintf("secret(s) %s misses resource %s among migrated resources", s.Name, strings.Join(missingStrings, ","))
+	}
+
+	return true, nil, ""
+}
+
+// keysWithPotentiallyPersistedData returns the minimal, recent secrets which have migrated all given GRs.
+func keysWithPotentiallyPersistedData(grs []schema.GroupResource, recentFirstSortedSecrets []*corev1.Secret) []*corev1.Secret {
```

Comment:
so this function actually`prunes/trims/removes/filters` keys that are no longer needed.

Reply:
yes, it takes the prefix of keys (starting with most recent) until it finds one which is migration key